services/galley/src/Galley/API/Federation.hs

```diff
@@ -115,22 +115,31 @@ onConversationUpdated requestingDomain cu = do
   localDomain <- viewFederationDomain
   let qconvId = Qualified (cuConvId cu) requestingDomain
 
+  -- Note: we generally do not send notifications to users that are not part of
+  -- the conversation (from our point of view), to prevent spam from the remote
+  -- backend. See also the comment below.
   (presentUsers, allUsersArePresent) <- Data.filterRemoteConvMembers (cuAlreadyPresentUsers cu) qconvId
 
-  -- perform action
+  -- Perform action, and determine extra notification targets.
+  --
+  -- When new users are being added to the conversation, we consider them as
+  -- notification targets. Once we start checking connections before letting
+  -- people being added, this will be safe against spam. However, if users that
+  -- are not in the conversations are being removed, we do **not** add them to the
+  -- list of targets, because we have no way to make sure that they are actually
+  -- supposed to receive that notification.
   extraTargets <- case cuAction cu of
     Public.ConversationActionAddMembers toAdd -> do
       let localUsers = getLocalUsers localDomain (fmap fst toAdd)
       Data.addLocalMembersToRemoteConv qconvId localUsers
       pure localUsers
     Public.ConversationActionRemoveMembers toRemove -> do
       let localUsers = getLocalUsers localDomain toRemove
-      -- TODO: do not remove users that are not in the conversation
       Data.removeLocalMembersFromRemoteConv qconvId localUsers
-      pure localUsers
+      pure []
```

Comment:
Interesting. Is this just to make dealing with spam easier? Will some legitimate users not be notified?

Reply:
Yes to both. See my explanation above.